### ISSUE TITLE
[Bug]: scatter `set_facecolor` does not work if the colors were manually set during the initial scatter call

### ISSUE THREAD

**Issue by scottshambaugh** [matplotlib/matplotlib#27555]
### Bug summary

If colors of scatter points are set during the initial call, then later calls of `set_facecolor` to that artist fail to update the colors. `set_edgecolor` works as expected. This is happening in 2D and 3D scatter plots (fixing the 2D case should fix both).

### Code for reproduction

```python
import matplotlib.pyplot as plt
import matplotlib as mpl
cmap = mpl.colormaps['viridis']
x = np.ones(10)
y = np.linspace(0, 1, 10)

fig, ax = plt.subplots()
scat = ax.scatter(x, y, c=y, cmap='viridis', s=20)
colors = cmap(np.roll(y, 3))
scat.set_edgecolor(colors)
scat.set_facecolor(colors)
```

### Actual outcome

![image](https://github.com/matplotlib/matplotlib/assets/14363975/e7f29167-3a98-462f-9d8c-b810abcd5a38)

### Expected outcome

Result if colors are not passed in the initial scatter call is as expected:
```python
fig, ax = plt.subplots()
scat = ax.scatter(x, y, s=20)
colors = cmap(np.roll(y, 3))
scat.set_edgecolor(colors)
scat.set_facecolor(colors)
```
![image](https://github.com/matplotlib/matplotlib/assets/14363975/a62fd62d-4e43-4724-aa0c-f3774414d18c)

### Matplotlib Version

latest main

### Installation

git checkout

**Comment by QuLogic** [comment]
I think this might be expected behaviour, but @efiring could confirm; colour mapping overrides face colour. To revert mapping, I think you need to call `scat.set_array(None)`.

**Comment by efiring** [comment]
I agree.  This is consistent with the docstring, and I think it is a reasonable choice of precedence.

**Comment by scottshambaugh** [comment]
Would it make sense to enforce the same behavior for edge colors? The inconsistency in behavior is rather confusing.

**Comment by efiring** [comment]
I don't understand.  How do you propose to change the edgecolor behavior?

**Comment by scottshambaugh** [comment]
Here's the docstring for scatter:

> carray-like or list of colors or color, optional
> The marker colors. Possible values:
> 
> A scalar or sequence of n numbers to be mapped to colors using cmap and norm.
> 
> A 2D array in which the rows are RGB or RGBA.
> 
> A sequence of colors of length n.
> 
> A single color format string.
> 
> Note that c should not be a single numeric RGB or RGBA sequence because that is indistinguishable from an array of values to be colormapped. If you want to specify the same RGB or RGBA value for all points, use a 2D array with a single row. Otherwise, value-matching will have precedence in case of a size matching with x and y.
> 
> If you wish to specify a single color for all points prefer the color keyword argument.
> 
> Defaults to [None](https://docs.python.org/3/library/constants.html#None). In that case the marker color is determined by the value of color, facecolor or facecolors. In case those are not specified or [None](https://docs.python.org/3/library/constants.html#None), the marker color is determined by the next color of the Axes' current "shape and fill" color cycle. This cycle defaults to [rcParams["axes.prop_cycle"]](https://matplotlib.org/stable/users/explain/customizing.html?highlight=axes.prop_cycle#matplotlibrc-sample) (default: cycler('color', ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf'])).

My read as a user is that this sets all the applicable colors (ie, both edge and face colors), since it seems like some of the available marker types are all edge and no face (eg `'+'`, though I'm not actually clear if that's true). Being able to later modify the edge colors but not face colors is confusing and seems inconsistent.

To push back a bit on the intended behavior here, I'm still not a fan of not being able to override the face colors after the function call - it's clear from the `scatter` docstring that the `facecolors` kwarg will be ignored if `c` is passed in during the initial call, but not clear that this is also the case if the user tries to `set_facecolor()` later on. If that's the intended behavior I suppose it's fine, but it's not clear in the `Collection.set_facecolor docstring`:

> Set the facecolor(s) of the collection. c can be a color (all patches have same color), or a sequence of colors; if it is a sequence the patches will cycle through the sequence.
> 
> If c is 'none', the patch will not be filled.

**Comment by paulbrodersen** [comment]
> I think it is a reasonable choice of precedence.

@efiring Could you expand on why you think that is the case? I am struggling to come up with scenarios where this would be the desired behaviour. 

As a user that just ran into this issue when trying to animate a scatter plot, this behaviour was very surprising, as from my perspective, it violates multiple assumptions that I have formed over the years when interacting with other matplotlib artists, and also seems counter in many respects to [PEP20](https://peps.python.org/pep-0020/):

1. Explicit is better than implicit, but here the explicit color assignment is ignored in favour of an implicit color assignment at instantiation. 
2. Special cases aren't special enough to break the rules, yet scatter behaves different than most (all?) other matplotlib artists in this respect. 
3. Errors should never pass silently, yet there is no feedback to the user that his or her color assignment via `set_facecolor` isn't doing anything. 

This is obviously just the perspective of a disgruntled user that spent way too long debugging and testing different backends (I was altering the alpha values via `set_facecolor` and thought it might be an opacity issue), ultimately refusing that matplotlib was to blame until all other options were exhausted. But I genuinely would like to understand if this is an instance of [Chesterton's fence](https://en.wiktionary.org/wiki/Chesterton%27s_fence), or just an artifact of the original spec.

**Comment by efiring** [comment]
Like much of matplotlib, the color handling has evolved, and if starting from scratch one could design a more explicit, consistent, and verbose API.  I think a problem with things like scatter is that a ScalarMappable has only one
array for mapping, but often 2 (edge and face) categories to which the mapping could be applied.  This has led
to the current approach in which there are precedence rules rather than more explicit instructions as to what
is to be mapped.  (The fact that some markers are implemented as edges only adds another source of complexity.)
Given that ScalarMappables exist to provide color-mapping, it is not unreasonable that they try (via the precedence
rules) to map the mappable array to something, provided that the array of numbers to be mapped is present.  Hence @QuLogic's comment above that if you have initially supplied an array, and then later you don't want anything to be mapped, you can call a method, `set_array(None)`.

Perhaps this method of disabling mapping should be explicitly added to the documentation?

In the current implementation via precedence rules, edgecolors and facecolors are not treated identically
because they are typically used in different ways. Edgecolors are expected to be used more frequently as 
outlines--for example, black outlines if the background is light, or white outlines if the
background is dark.  Facecolors are expected to be the primary targets of color mapping.

**Comment by timhoffm** [comment]
> Errors should never pass silently, yet there is no feedback to the user that his or her color assignment via set_facecolor isn't doing anything.

Indeed. This should issue a warning if the ScalarMappable has array data to make the user aware of the mapping precedence and that set facecolor likely does not do what the user expects.

**Comment by dpjensen** [comment]
I agree with timhoffm, even if the behavior doesn't change attempting to set facecolor when an array is set for the color should throw a warning. Especially because this behavior did not exist for older versions of matplotlib, set_color used to override any previous color for both the edge and face colors. I just updated my python environment and spent an entire day trying to figure out why my plotting code was no longer updating the color before I narrowed the issue down to the facecolor getting ignored and found this bug report. Calling set_array(None) is an easy enough fix, but only if users know they need to do that, and currently this bug report seems to be the only place that is documented.

**Comment by jhkob** [comment]
Hi, I just spent several hours fighting this very behavior, since I want to use a different face color for the markers in legend. (I use edge color to mark a few classes of data, face color for a variable with cmap.)

In an attempt to engineer the legend, I created a copy of the handles and tried to change the facecolor in the legend entries. Perhaps I am missing something, but it seems this is just impossible to do now. Or is there a method e.g. to re-create the markers just for legend so they would inherit all properties except for the cmap ?  This would really be needed for my use case...
